grpcbp/server_middlewares.go

```diff
@@ -0,0 +1,94 @@
+package grpcbp
+
+import (
+	"context"
+	"strings"
+
+	"github.com/reddit/baseplate.go/ecinterface"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/reddit/baseplate.go/transport"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+func InjectServerSpanInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		split := strings.Split(info.FullMethod, "/")
+		m := split[len(split)-1]
+		ctx, span := StartSpanFromGRPCContext(ctx, m)
+
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			if values := md.Get(transport.HeaderTracingTrace); len(values) > 0 {
+				span.SetTag(tracing.TagKeyPeerService, values[0])
+			}
+		}
+
+		defer func() {
+			span.FinishWithOptions(tracing.FinishOptions{
+				Ctx: ctx,
+				Err: err,
+			}.Convert())
+		}()
+		return handler(ctx, req)
+	}
+}
+
+func InjectEdgeContextInterceptor(impl ecinterface.Interface) grpc.UnaryServerInterceptor {
+	if impl == nil {
+		impl = ecinterface.Get()
+	}
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (_ interface{}, err error) {
+		ctx = InitializeEdgeContext(ctx, impl)
+		return handler(ctx, req)
+	}
+}
+
+func InitializeEdgeContext(ctx context.Context, impl ecinterface.Interface) context.Context {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return ctx
+	}
+	values := md.Get(transport.HeaderEdgeRequest)
+	if len(values) == 0 {
+		return ctx
+	}
+
+	ctx, err := impl.HeaderToContext(ctx, values[0])
+	if err != nil {
+		log.C(ctx).Errorw("Error while parsing EdgeRequestContext: " + err.Error())
+	}
+	return ctx
+}
+
+func StartSpanFromGRPCContext(ctx context.Context, name string) (context.Context, *tracing.Span) {
+	var (
+		headers tracing.Headers
+		sampled bool
+	)
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return tracing.StartSpanFromHeaders(ctx, name, headers)
+	}
```

Comment:
same as above, I don't think you need to care much about the `!ok` case (on the read side)

Reply:
That would mean the method exits silently though without creating a span.